mesonbuild/backend/xcodebackend.py

```diff
@@ -1052,6 +1074,24 @@ def generate_pbx_group(self, objects_dict: PbxDict) -> None:
         main_children.add_item(frameworks_id, 'Frameworks')
         main_dict.add_item('sourceTree', '<group>')
 
+        # Define each folder as a group in Xcode. That way, it can build the file tree correctly.
+        # This must be done before the project tree group is generated, as source files are added during that phase.
+        for (path, target), id in self.foldermap.items():
+            folder_dict = PbxDict()
+            objects_dict.add_item(id, folder_dict, path)
+            folder_dict.add_item('isa', 'PBXGroup')
+            folder_children = PbxArray()
+            folder_dict.add_item('children', folder_children)
+            folder_dict.add_item('name', '"%s"' % path.split('/')[-1])
+            folder_dict.add_item('path', '"%s"' % path)
+            folder_dict.add_item('sourceTree', 'SOURCE_ROOT')
+
+            # Add any detected subdirectories (not declared as subdir()) here, but only one level higher.
+            # Example: In "root", add "root/sub", but not "root/sub/subtwo".
+            for path_dep, target_dep in self.foldermap:
+                if path_dep.startswith(path) and path_dep.split('/')[0] == path.split('/')[0] and path_dep != path and len(path_dep.split('/')) == len(path.split('/')) + 1 and target == target_dep:
```

Comment:
At the very least we should do this for a bit of optimization:
```suggestion
                if path_dep.startswith(path) and path_dep.split('/', 1)[0] == path.split('/', 1)[0] and path_dep != path and path_dep.count('/') == path.count('/') + 1 and target == target_dep:
```

I still think it would be better to use the `os.path` methods here for clarity, and because they can handle things that string manipulation might miss (like a path ending in `/` being counted incorrectly, which will happen with both the `.split()` and `.count()` approaches).